repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/BadRow.scala

```diff
@@ -0,0 +1,177 @@
+
+import java.util.{List => JList, Map => JMap}
+
+import scala.collection.JavaConverters._
+import cats.data.NonEmptyList
+import io.circe.syntax._
+import io.circe.{Encoder, Json}
+import io.circe.literal._
+import com.google.cloud.bigquery.{BigQueryException, BigQueryError => JBigQueryError}
+import com.snowplowanalytics.iglu.core.SelfDescribingData
+import com.snowplowanalytics.iglu.core.circe.instances._
+import com.snowplowanalytics.snowplow.storage.bigquery.common.{BadRowSchemas, ProcessorInfo}
+import com.snowplowanalytics.snowplow.storage.bigquery.repeater.PayloadParser.ReconstructedEvent
+import com.snowplowanalytics.snowplow.storage.bigquery.repeater.generated.ProjectMetadata
+
+
+/**
+  * Represents rows of failed inserts which are problematic
+  * to process due to several reasons such as could not be
+  * converted back to enrich event, error while trying to
+  * insert row to BigQuery etc
+  */
+sealed trait BadRow {
+  def getSelfDescribingData: SelfDescribingData[Json]
+  def compact: String  = getSelfDescribingData.asJson.noSpaces
+}
+
+object BadRow {
+  val BQLoaderProcessorInfo = ProcessorInfo.Processor(ProjectMetadata.name, ProjectMetadata.version)
+
+  /**
+    * Represents situations where payload object can not be
+    * converted back to enriched event format successfully
+    * @param payload data which is tried to be converted back to enrich event
+    * @param errors gives info about reasons of failure to convert it back
+    */
+  final case class ParsingError(payload: String, errors: NonEmptyList[ParsingErrorInfo]) extends BadRow {
+    def getSelfDescribingData: SelfDescribingData[Json] =
+      SelfDescribingData(BadRowSchemas.RepeaterParsingError, (this: BadRow).asJson)
+  }
+
+  /**
+    * Gives info about reasons of parsing error
+    */
+  sealed trait ParsingErrorInfo
+
+  /**
+    * Gives info about reasons of failure to parse payload to event
+    * @param message error message
+    * @param location location in the JSON object where error happened
+    */
+  final case class EventParsingError(message: String, location: List[String]) extends ParsingErrorInfo
+
+  /**
+    * Gives info about reasons of failure to parse payload to json
+    * @param error error message
+    */
+  final case class JsonParsingError(error: String) extends ParsingErrorInfo
+
+  implicit val parsingErrorInfoEncoder: Encoder[ParsingErrorInfo] =
+    Encoder.instance {
+      case EventParsingError(message, location) => json"""{
+          "eventParsingError": {
+            "error": $message,
+            "location": $location
+          }
+        }"""
+      case JsonParsingError(error) => json"""{
+          "jsonParsingError": {
+            "error": $error
+          }
+        }"""
+    }
+
+  /**
+    * Represents situations where something went wrong internally
+    * In this kind of situations, payload data can be successfully
+    * converted back to enrich event format but some internal error
```

Comment:
What is "internal error"?

Reply:
Will remove the `internal` keyword as I mentioned before. Will make comment more detailed.

Reply:
Just in case. Its not that much about level of details, but rather about giving a new context or point of view, or just explaning things in entirely different terms.